crates/services/p2p/src/discovery.rs

```diff
@@ -184,51 +154,20 @@ impl NetworkBehaviour for DiscoveryBehaviour {
             }
         }
 
-        // poll Kademlia behaviour
-        while let Poll::Ready(kad_action) = self.kademlia.poll(cx, params) {
+        // poll sub-behaviors
+        if let Poll::Ready(kad_action) = self.kademlia.poll(cx, params) {
             match kad_action {
-                NetworkBehaviourAction::GenerateEvent(
-                    KademliaEvent::UnroutablePeer { peer },
-                ) => {
-                    return Poll::Ready(NetworkBehaviourAction::GenerateEvent(
-                        DiscoveryEvent::UnroutablePeer(peer),
-                    ))
-                }
-
-                NetworkBehaviourAction::Dial { handler, opts } => {
-                    return Poll::Ready(NetworkBehaviourAction::Dial { handler, opts })
-                }
-                NetworkBehaviourAction::CloseConnection {
-                    peer_id,
-                    connection,
-                } => {
-                    return Poll::Ready(NetworkBehaviourAction::CloseConnection {
-                        peer_id,
-                        connection,
-                    })
-                }
-                NetworkBehaviourAction::NotifyHandler {
-                    peer_id,
-                    handler,
-                    event,
-                } => {
-                    return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
-                        peer_id,
-                        handler,
-                        event,
-                    })
-                }
-                NetworkBehaviourAction::ReportObservedAddr { address, score } => {
-                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
-                        address,
-                        score,
-                    })
-                }
-                _ => {}
+                NetworkBehaviourAction::GenerateEvent(event) => match event {
+                    KademliaEvent::UnroutablePeer { peer } => {
+                        Poll::Ready(NetworkBehaviourAction::GenerateEvent(
+                            KademliaEvent::UnroutablePeer { peer },
+                        ))
+                    }
+                    _ => Poll::Pending,
+                },
+                _ => Poll::Ready(kad_action),
             }
-        }
-
-        while let Poll::Ready(mdns_event) = self.mdns.poll(cx, params) {
+        } else if let Poll::Ready(mdns_event) = self.mdns.poll(cx, params) {
```

Comment:
Hmm, the previous iteration would iterate over all mdns events. Let's say, if we had many `NetworkBehaviourAction::GenerateEvent(MdnsEvent::Discovered(list))`, we would process them for one loop.

While right not it is not true. I'm okay to use `if` statement for Kademlia, but for mdns it is better to use `while` as we did before=)